indexer/eventprocessor.go

```diff
@@ -2,69 +2,100 @@ package indexer
 
 import (
 	"context"
+	"sync"
 	"time"
 
 	"gorm.io/gorm"
 
 	"github.com/photon-storage/go-common/log"
+	"github.com/photon-storage/go-photon/crypto/sha256"
 
 	"github.com/photon-storage/photon-explorer/chain"
 	"github.com/photon-storage/photon-explorer/database/orm"
 )
 
 // EventProcessor is the processor for synchronizing photon change events.
 type EventProcessor struct {
+	ctx             context.Context
 	refreshInterval uint64
 	db              *gorm.DB
 	node            *chain.NodeClient
 	currentSlot     uint64
 	currentHash     string
+	quit            chan struct{}
+	wg              sync.WaitGroup
 }
 
 // NewEventProcessor returns the new instance of EventProcessor.
-func NewEventProcessor(refreshInterval uint64, nodeEndpoint string, db *gorm.DB) *EventProcessor {
+func NewEventProcessor(
+	ctx context.Context,
+	refreshInterval uint64,
+	nodeEndpoint string,
+	db *gorm.DB,
+) *EventProcessor {
 	currentSlot, currentHash, err := currentChainStatus(db)
 	if err != nil {
 		log.Fatal("fail query chain status", "error", err)
 	}
 
 	return &EventProcessor{
+		ctx:             ctx,
 		refreshInterval: refreshInterval,
 		db:              db,
 		node:            chain.NewNodeClient(nodeEndpoint),
 		currentSlot:     currentSlot,
 		currentHash:     currentHash,
+		quit:            make(chan struct{}),
 	}
 }
 
+// Stop exit event processor
+func (e *EventProcessor) Stop() {
+	if err := e.db.Model(&orm.ChainStatus{}).Where("id = 1").Updates(
+		map[string]interface{}{
+			"slot": e.currentSlot,
+			"hash": e.currentHash,
+		}).Error; err != nil {
+		log.Fatal(err.Error())
+	}
+	close(e.quit)
+	e.wg.Wait()
+}
+
 // Run executing the timing task of processing chain data.
-func (e *EventProcessor) Run(ctx context.Context) {
+func (e *EventProcessor) Run() {
 	ticker := time.NewTicker(time.Duration(e.refreshInterval) * time.Second)
 	defer ticker.Stop()
 
-	for {
+	e.wg.Add(1)
+	for ; true; <-ticker.C {
```

Comment:
this implementation is not ideal. When it waits on ticker.C, it won't process e.quit and e.ctx.Done() signal